src/browser/provider/built-in/saucelabs.js

```diff
@@ -0,0 +1,317 @@
+                    platformGroup
+                        .list
+                        .map(device => this._getDeviceInfo(device, {
+                            automationApi,
+                            platformGroup: platformGroup.name
+                        }))
+                ))
+        );
+    },
+
+    async _fetchPlatformInfoAndAliases () {
+        var platformsTree = await Promise.all(
+            AUTOMATION_APIS.map(automationApi => this._getAutomationApiInfo(automationApi))
+        );
+
+        this.platformsInfo = flattenDeep(platformsTree);
+
+        var unstructuredBrowserNames = this.platformsInfo
+            .map(platformInfo => this._createAliasesForPlatformInfo(platformInfo));
+
+        this.availableBrowserNames = flatten(unstructuredBrowserNames);
+    },
+
+    _createAliasesForPlatformInfo (platformInfo) {
+        if (platformInfo.device === 'Android Emulator') {
+            return [
+                this._createAliasesForPlatformInfo(assign({}, platformInfo, { device: 'Android Emulator Tablet' })),
+                this._createAliasesForPlatformInfo(assign({}, platformInfo, { device: 'Android Emulator Phone' }))
+            ];
+        }
+
+        var name     = isSelenium(platformInfo) ? platformInfo.browserName : platformInfo.device;
+        var version  = isSelenium(platformInfo) ? platformInfo.browserVersion : platformInfo.os;
+        var platform = isSelenium(platformInfo) ? platformInfo['os'] : '';
+
+        if (platformInfo.automationApi === 'appium' && platformInfo.platformGroup === 'Android')
+            name += ' Appium';
+
+        return `${name}@${version}${platform ? ':' + platform : ''}`;
+    },
+
+    _createQuery (browserName) {
+        var { browserName: name, browserVersion: version, platform } = parseCapabilities(browserName)[0];
+
+        var query = { name, version, platform };
+
+        if (/appium$/.test(name)) {
+            query.useAppium = true;
+            query.name      = query.name.replace(' appium', '');
+        }
+
+        if (/^android emulator/.test(name)) {
+            query.deviceType = query.name.replace('android emulator ', '');
+            query.name       = 'android emulator';
+        }
+
+        return query;
+    },
+
+    _filterPlatformInfo (query) {
+        return this.platformsInfo
+            .filter(info => {
+                return info.browserName && info.browserName.toLowerCase() === query.name ||
+                       info.device && info.device.toLowerCase() === query.name;
+            })
+            .filter(info => {
+                return info.browserVersion === query.version || info.browserVersion === query.version + '.0' ||
+                       info.os === query.version || query.version === 'any';
+            })
+            .filter(info => info.os === query.platform || query.platform === 'any')
+            .filter(info => info.automationApi !== 'appium' || info.platformGroup !== 'Android' || query.useAppium);
```

Comment:
It's better have one filter here, with vars extracted. Because I still have no idea what each of these four filters trying to filter.